influxdb/metric.go

```diff
@@ -0,0 +1,224 @@
+}
+
+// MetricConfig defines the configuration file parameters for a sepcific metric
+// imported from InfluxDB.
+type MetricConfig struct {
+	Query    string
+	Database string
+	Endpoint string
+	Username string
+	Password string
+}
+
+func NewSourceMetric(name string, config *MetricConfig, offsetDuration, counterResetInterval time.Duration) (*Metric, error) {
+	return &Metric{
+		Name:                 name,
+		config:               config,
+		offsetDuration:       offsetDuration,
+		counterResetInterval: counterResetInterval,
+	}, nil
+}
+
+func (m *Metric) StackdriverName() string {
+	return fmt.Sprintf("custom.googleapis.com/influxdb/%s", m.Name)
+}
+
+func (m *Metric) Query() string {
+	return m.config.Query
+}
+
+func (m *Metric) StackdriverData(ctx context.Context, lastPoint time.Time, _ record.MetricRecord) (*metricpb.MetricDescriptor, []*monitoringpb.TimeSeries, error) {
+	c, err := client.NewHTTPClient(client.HTTPConfig{
+		Addr:     m.config.Endpoint,
+		Username: m.config.Username,
+		Password: m.config.Password,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create InfluxDB client: %v", err)
+	}
+
+	defer c.Close()
+
+	// We query from [startTime, endTime), where startTime is the timestamp
+	// of the latest point plus a nanosecond offset as it is inclusive, and
+	// endTime is the current time with an offset back as points that are too
+	// fresh may contain incomplete data.
+	startTime := lastPoint.Add(time.Nanosecond)
+	endTime := timeNow().Add(-m.offsetDuration)
+
+	resp, err := c.Query(m.buildQuery(startTime, endTime))
+	if err != nil {
+		return nil, nil, err
+	} else if err = resp.Error(); err != nil {
+		return nil, nil, err
+	}
+
+	if len(resp.Results) != 1 {
+		return nil, nil, fmt.Errorf("InfluxDB query '%s' returned %d query results, expected 1", m.config.Query, len(resp.Results))
+	}
+
+	if len(resp.Results[0].Series) == 0 {
+		log.Infof(ctx, "InfluxDB query '%s' returned no time series", m.config.Query)
+		return nil, nil, nil
+	} else if len(resp.Results[0].Series) > 1 {
+		return nil, nil, fmt.Errorf("InfluxDB query '%s' returned %d time series", m.config.Query, len(resp.Results[0].Series))
+	}
+
+	serie := resp.Results[0].Series[0]
+	points, err := parseSeriePoints(serie)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to parse InfluxDB points: %v", err)
+	}
+
+	// For gauge metrics, we don't need to apply any filtering since we only
+	// query InfluxDB for new points, and points that are too fresh are ignored
+	// by applying an offset. We'll need to apply filtering once cumulative
+	// metrics come into play.
+	return m.metricDescriptor(), m.convertTimeSeries(points), nil
+}
+
+// buildQuery creates a InfluxDB query from the metric query definition,
+// wrapped in the given time-interval, inclusive of start time.
+func (m *Metric) buildQuery(startTime, endTime time.Time) client.Query {
+	query := fmt.Sprintf(
+		"SELECT * FROM (%s) WHERE time >= %s AND time < %s",
```

Comment:
I think you should be able to use `%d` rather than `%s` for integer parameters here, instead of converting them to strings using `FormatInt`.

Reply:
whoops good catch, done